### PR TITLE
Addtional invoice fields

### LINKED ISSUE #367: Lower level CLN invoice commands
When trying to implement features that go beyond a simple wallet, one hits roadblocks pretty quickly. I was trying to implement [LNURL for the Breez SDK](https://github.com/michael1011/clnurl/pull/1), but unfortunately, that is not possible because the `description_hash` of the invoices from the SDK can't be set.

I would propose exposing more fields of [the `create_invoice` function](https://github.com/breez/breez-sdk/blob/main/libs/sdk-core/src/greenlight/node_api.rs#L242).
Or exposing "dev" commands like `signinvoice` that would allow for getting an invoice with the `create_invoice` call, decoding, setting your own fields and signing again.

### LINKED ISSUE #404: Feature request: allow `deschashonly=True` for CLN invoices
Please allow setting the flag `deschashonly=True` via the SDK so that we can create LNURL invoices.

### PR DESCRIPTION
**Pull request by roeierez**
In this PR we add additional optional invoice fields.
In addition uniffi default parameters are used for the first time. Based on the feedback, if agreed, we will apply the defaults for other optional fields to simplify the user experience.
fixed #404 
fixed #367

### CODE REVIEW

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt [hunk 1]
```diff
@@ -110,8 +110,10 @@ fun asReceivePaymentRequest(reqData: ReadableMap): ReceivePaymentRequest? {
     if (description != null) {
         val preimage = reqData.getArray("preimage")?.let { asUByteList(it) }
         val openingFeeParams = reqData.getMap("openingFeeParams")?.let{ asOpeningFeeParams(it) }
-
-        return ReceivePaymentRequest(amountSats.toULong(), description, preimage, openingFeeParams)
+        val useDescriptionHash : Boolean? = if (hasNonNullKey(reqData, "useDescriptionHash")) reqData.getBoolean("useDescriptionHash") else null;
+        val expiry : ULong? = if (hasNonNullKey(reqData, "expiry")) reqData.getInt("expiry").toULong() else null;
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt:114]: This `getInt` limits the value range to 32 bits, I think its better to use `getDouble` when converting to Long/ULong.

**Review by dangeross — CHANGES_REQUESTED** [review]

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt [hunk 2]
```diff
@@ -110,8 +110,10 @@ fun asReceivePaymentRequest(reqData: ReadableMap): ReceivePaymentRequest? {
...
+        val cltv : UInt? = if (hasNonNullKey(reqData, "expiry")) reqData.getInt("cltv").toUInt() else null;
```
[the remark this answers is not in the document]
Comment (roeierez) [reply libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt]: Nice catch! done.

File: libs/sdk-bindings/src/breez_sdk.udl [hunk 3]
```diff
@@ -295,8 +295,11 @@ dictionary ReverseSwapFeesRequest {
 dictionary ReceivePaymentRequest {
     u64 amount_sats;
     string description;
-    sequence<u8>? preimage;
-    OpeningFeeParams? opening_fee_params;
+    sequence<u8>? preimage = null;
+    OpeningFeeParams? opening_fee_params = null;
+    boolean? use_description_hash = null;
```
[the remark this answers is not in the document]
Comment (roeierez) [reply libs/sdk-bindings/src/breez_sdk.udl:300]: It turns out that without the default fields the uniffi generates code that still requires the user to pass them in some languages like python (even if the user just passes None, still required).
With the default fields initialized the use is not required to address them in the request.
As for the value, null means use default (of the underline c-lightning API which is taken as false at the end) while false means explicitly false is sent to greenlight.

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt (hunk 1 above)
Comment (roeierez) [reply libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt:114]: But we are not converting to Long/ULong we are converting to UInt.

Comment (dangeross) [reply libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt:114]: `expiry` is u64, so needs `toULong()`

File: libs/sdk-bindings/src/breez_sdk.udl (hunk 3 above)
[the remark this answers is not in the document]
Comment (dangeross) [reply libs/sdk-bindings/src/breez_sdk.udl:300]: Makes sense

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt (hunk 1 above)
Comment (roeierez) [reply libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt]: Right, I confused that with cltv. Fixed.

**Review by dangeross — APPROVED** [review]
LGTM!

**Comment by ok300** [comment]
I created an invoice[^1] without the extra flags

```
receive_payment 5000 "test-recv-plain"

// bolt11: lnbc50u1pjw4l2qsp59mw24ptul8943d5s3m7y0azxmwnc4wh2ptvvk6u6fak9k334hngqpp5wv47mnzqyu43cung9ymdjg9ytp0yujqz9gurfpmj2e93wqh0wnqsdquyf6x2um594ex2cmk94cxcctfdc3qxqyjw5qcqpxrzjqtypret4hcklglvtfrdt85l3exc0dctdp4qttmtcy5es3lpt6uts6vgenmxque8jkgqqqqlgqqqqqzsqyg9qxpqysgqtasvxy64f3exajgxnwdzfcvvh49m4dsl7vz4jznlcgqq87dfgl8knsclm3cuvjln50rdv52szngx7texx9m0g2uwwsk5auvxhev37mcpj0t2cu
```

and one[^2] with extra flags

```
receive_payment -c 100 -e 1000 -s true 5500 "test-recv-pr"

// bolt11: lnbc55u1pjw4ld4sp5wnmylkthz64km784nrx7lew5hx0pzv99xr8mtckv29g82yc2e74spp56nkx97jtkvf2nsdj54ul3x7ev9988ajpqut0nuaax4z99l7snunqdqhyf6x2um594ex2cmk94c8ygsxqyjw5qcqpxrzjqtypret4hcklglvtfrdt85l3exc0dctdp4qttmtcy5es3lpt6uts6vgenmxque8jkgqqqqlgqqqqqzsqyg9qxpqysgqg64cl0670dx3e5sf6t7z44kjxfw42t8tzawlxzpkz3xku856truhxh6rr5pd3ttrs4mmmp0zntepmhkez50j6tety7h2kxq52k2ey3qqf4zxay
```

They both ended up having the same expire time of 604800, despite the flag `-e 1000` in the 2nd command.

Am I doing something wrong? Shouldn't the 2nd expiry be different?



[^1]: Decoded: https://lightningdecoder.com/lnbc50u1pjw4l2qsp59mw24ptul8943d5s3m7y0azxmwnc4wh2ptvvk6u6fak9k334hngqpp5wv47mnzqyu43cung9ymdjg9ytp0yujqz9gurfpmj2e93wqh0wnqsdquyf6x2um594ex2cmk94cxcctfdc3qxqyjw5qcqpxrzjqtypret4hcklglvtfrdt85l3exc0dctdp4qttmtcy5es3lpt6uts6vgenmxque8jkgqqqqlgqqqqqzsqyg9qxpqysgqtasvxy64f3exajgxnwdzfcvvh49m4dsl7vz4jznlcgqq87dfgl8knsclm3cuvjln50rdv52szngx7texx9m0g2uwwsk5auvxhev37mcpj0t2cu

[^2]: Decoded: https://lightningdecoder.com/lnbc55u1pjw4ld4sp5wnmylkthz64km784nrx7lew5hx0pzv99xr8mtckv29g82yc2e74spp56nkx97jtkvf2nsdj54ul3x7ev9988ajpqut0nuaax4z99l7snunqdqhyf6x2um594ex2cmk94c8ygsxqyjw5qcqpxrzjqtypret4hcklglvtfrdt85l3exc0dctdp4qttmtcy5es3lpt6uts6vgenmxque8jkgqqqqlgqqqqqzsqyg9qxpqysgqg64cl0670dx3e5sf6t7z44kjxfw42t8tzawlxzpkz3xku856truhxh6rr5pd3ttrs4mmmp0zntepmhkez50j6tety7h2kxq52k2ey3qqf4zxay

**Comment by roeierez** [comment]
@ok300 thanks for testing!
You are not doing anything wrong, there is an issue in greenlight for these fields. I communicated that to Christian. Once it is fixed on gl side it should work.

**Review by ok300 — APPROVED** [review]
LGTM!

The moonpay test looks unrelated to this PR, so is the GL handling of expiry.

File: libs/sdk-core/src/models.rs [hunk 4]
```diff
@@ -667,6 +670,9 @@ pub struct ReceivePaymentRequest {
     pub description: String,
     pub preimage: Option<Vec<u8>>,
     pub opening_fee_params: Option<OpeningFeeParams>,
+    pub use_description_hash: Option<bool>,
+    pub expiry: Option<u64>,
+    pub cltv: Option<u32>,
```
[the remark this answers is not in the document]
Comment (roeierez) [reply libs/sdk-core/src/models.rs:684]: Right, done.

**Review by JssDWt — APPROVED** [review]

**Pull request merged**